### PR TITLE
feature/auto-build-db-diagram

### PR DESCRIPTION
**Pull request by JacksonMaxfield**
**Pull request recommendations:**
- [x] Name your pull request _your-development-type/short-description_. Ex: _feature/read-tiff-files_
- [x] Link to any relevant issue in the PR description. Ex: _Resolves [gh-12], adds tiff file format support_
- [x] Provide context of changes.

I wanted to write a script that auto-checks for cyclic dependencies in our database schema as I think I see some coming up in future PRs / model additions. I quickly realized that this could be handled for us by many existing systems that handle directed graphs / graph builds and visualizers and figured I could two-for-one, generate a db diagram and also check for cyclic dependencies in one go with one such tools.

I first tried with [`pyreverse`](https://www.logilab.org/blogentry/6883) from [`pylint`](https://www.pylint.org/) but found it wasn't flexible enough. I noticed however that it (much like everything else in the DAG world) utilizes [`dot` / `graphviz`](https://graphviz.org/) under-the-hood and figured I would just write my own custom handler.

So that's what I did.

We now have a db diagram that is built as part of `make docs` (and `make gen-docs`).

image

Let me know if you want styling changed and such, it's finicky but possible.

- [x] Provide relevant tests for your feature or bug fix.
- [x] Provide or update documentation for any feature added by your pull request.

Thanks for contributing!

### CODE REVIEW

File: cdp_backend/bin/create_cdp_database_uml.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+            "bgcolor": "transparent",
+            "splines": "compound",
+        },
+        node_attr={
+            "shape": "record",
+            "style": "filled",
+            "fillcolor": "white",
+        },
+        edge_attr={
+            "arrowhead": "none",
```
Comment (tohuynh) [inline cdp_backend/bin/create_cdp_database_uml.py:63]: I think adding an arrowhead `vee` or `normal` would add some visual cues to help us comprehend the diagram easier. So we don't have check whether the line is attaching to rectangle as a whole or to a particular field.

File: cdp_backend/bin/create_cdp_database_uml.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            node_label = f"{model_name} | {fields_as_text}"
+
+            # Attach as a complete node
+            dot.node(model_name, node_label)
+
+    # Second pass: Create DAG
+    for model_name, cls in inspect.getmembers(models, inspect.isclass):
+        if model_name not in ["Model", "datetime"]:
+            # Attach fields for each model by using the Example
+            fields = []
```
Comment (tohuynh) [inline cdp_backend/bin/create_cdp_database_uml.py:111]: This line is not needed?

**Review by tohuynh — APPROVED** [review]
This is really cool. Was it your intent to create the diagram and then manually inspect for cycles, or is there suppose to be code in create_cdp_database_uml.py to automatically detect cycles?

File: cdp_backend/bin/create_cdp_database_uml.py (hunk 2 above)
Comment (JacksonMaxfield) [reply cdp_backend/bin/create_cdp_database_uml.py:111]: Good catch!

File: cdp_backend/bin/create_cdp_database_uml.py (hunk 1 above)
Comment (JacksonMaxfield) [reply cdp_backend/bin/create_cdp_database_uml.py:63]: Yea I debated this for a bit. I can change back `normal`. Makes sense.

**Review by JacksonMaxfield — COMMENTED** [review]
Good catch on the `Digraph` instead of `DAG`. Will add in a step to check for cyclic deps.

File: cdp_backend/bin/create_cdp_database_uml.py (hunk 1 above)
Comment (JacksonMaxfield) [reply cdp_backend/bin/create_cdp_database_uml.py]: Changed to `vee`

**Review by isaacna — APPROVED** [review]

**Pull request merged**